leshan-core/src/main/java/org/eclipse/leshan/core/request/argument/Arguments.java

```diff
@@ -0,0 +1,331 @@
+
+        builder.addArgument(digit, value);
+        return currentCursor;
+    }
+
+    /**
+     * Serialize {$link Arguments} into text representation. In case of empty arguments returns null.
+     */
+    public String serialize() {
+        StringBuilder sb = new StringBuilder();
+
+        for (Argument argument : getValues()) {
+            if (sb.length() > 0) {
+                sb.append(",");
+            }
+
+            sb.append(argument.getDigit());
+
+            if (argument.getValue() != null) {
+                sb.append("='");
+                sb.append(argument.getValue());
+                sb.append("'");
+            }
+        }
+
+        if (sb.toString().length() == 0) {
+            return null;
+        }
+
+        return sb.toString();
+    }
+
+    /**
+     * Creates a map that represents arguments.
+     */
+    public Map<Integer, String> toMap() {
+        Map<Integer, String> result = new HashMap<>();
+        for (Argument argument : getValues()) {
+            String value = argument.getValue();
+            result.put(argument.getDigit(), value);
+        }
+        return result;
+    }
+
+    /**
+     * Create a builder for {$link Arguments}.
+     */
+    public static ArgumentsBuilder builder() {
+        return new ArgumentsBuilder();
+    }
+
+    /**
+     * Builder for creating {$link Arguments}.
+     */
+    public static class ArgumentsBuilder {
+
+        private final List<PendingArgument> pendingArguments = new ArrayList<>();
+
+        /**
+         * Add an argument with digit and value.
+         */
+        public ArgumentsBuilder addArgument(int digit, String value) {
+            pendingArguments.add(new PendingArgument(digit, value));
+            return this;
+        }
+
+        /**
+         * Add an argument with digit and no value.
+         */
+        public ArgumentsBuilder addArgument(int digit) {
+            pendingArguments.add(new PendingArgument(digit, null));
+            return this;
+        }
+
+        /**
+         * Add an argument.
+         */
+        public void addArgument(Argument argument) {
+            if (argument != null) {
+                addArgument(argument.getDigit(), argument.getValue());
+            }
+        }
+
+        /**
+         * Builds {$link Arguments}.
+         *
+         * @throws InvalidArgumentException In case of any argument is invalid or digit exists more than once.
+         */
+        public Arguments build() throws InvalidArgumentException {
+            Map<Integer, Argument> argumentMap = new LinkedHashMap<>();
+
+            for (PendingArgument argument : pendingArguments) {
+                if (argumentMap.containsKey(argument.digit)) {
+                    throw new InvalidArgumentException("Unable to build Arguments : digit [%d] exists more than once",
+                            argument.digit);
+                }
+                argumentMap.put(argument.digit, new Argument(argument.digit, argument.value));
```

Comment:
I'm not sure if we should raise this exception on `build` like we currently do.
OR if we should raise the exception as soon as possible on `addArgument` ?

The second way seems to be better to because code will be simpler and we failed as soon as possible.

Reply:
In my opinion in general failing on build() method instead of parameters method is better, because build() method has more context for validation. In ArgumentsBuilder it seems no matter, but to be more coherent with potential future builders it's better to make validation in build() method.

There is discussion about when to fail: https://softwareengineering.stackexchange.com/questions/241309/builder-pattern-when-to-fail

Reply:
I read the discussion and I get all the pros&cons makes senses and this still hard to be to know which move in the best one.

I guess we can keep it like this.